📄 Update Japanese OAuth docs (#1697)
slackapi/bolt-js: docs/_basic/ja_authenticating_oauth.md

---

<div class="section-content">
Slack アプリの配布準備のために、 Bolt による OAuth フローを実装し、インストール時に取得した情報をセキュアな方法で保存しておく必要があります。
Bolt は OAuth による認証そのものと、OAuth ルーティング、 state にる認証、保存すべきインストール情報の受け渡しなどの周辺機能もサポートしています。

OAuth を有効にするために必要なもの:
* `clientId`, `clientSecret`, `stateSecret`, `scopes` _(必須)_
* `installationStore` は、データベースへのアプリ情報の読み込みと書き込みをするハンドラーのオプションです *(任意、プロダクションコードにおいては強く推奨)*

##### 開発とテスト

開発とテストにおいて利用可能な `installationStore` オプションの実装 `FileInstallationStore` を Bolt は既に用意しています。

```javascript
const { App } = require('@slack/bolt');
const { FileInstallationStore } = require('@slack/oauth');
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  clientId: process.env.SLACK_CLIENT_ID,
  clientSecret: process.env.SLACK_CLIENT_SECRET,
  stateSecret: 'my-state-secret',
  scopes: ['channels:history', 'chat:write', 'commands'],
  installationStore: new FileInstallationStore(),
});
```
:warning: プロダクションコードとしての利用は **推奨しません** ので、自前で実装する必要があります。サンプルコードとして [OAuth の別の実装例](https://github.com/slackapi/bolt-js/tree/main/examples/oauth)を参照してください。

##### アプリのインストール

> * **インストール時の初期化**: Bolt for JavaScript は `slack/install` という **パス** を生成します。これは、Slack アプリのダイレクトインストールのために `Add to Slack` ボタンを置く場合に指定できる URL です（ `state` パラメータが必要）。例えば、 _www.example.com_ でホスティングされているアプリであれば、 _www.example.com/slack/install_ とインストールページが提供されます。

Reviewer: ```suggestion
* **インストールの開始**: Bolt for JavaScript は `/slack/install` という **インストール用のパス** を生成します。これは、有効な `state` パラメータを生成した上で Slack アプリの直接のインストールを開始するための `Add to Slack` ボタンを含むページを応答する URL です。 _www.example.com_ でホスティングされているアプリの場合、インストールページは _www.example.com/slack/install_ となります。
```

>   * 💡 `App` コンストラクタ内で `installerOptions.directInstall: true` と設定することで、デフォルトのウェブページを描画しない代わりに、ユーザーへ Slack の認証 URL を出すことができます（[例](https://github.com/slackapi/bolt-js/blob/5b4d9ceb65e6bf5cf29dfa58268ea248e5466bfb/examples/oauth/app.js#L58-L64)）。

Reviewer: ```suggestion
  * 💡 `App` コンストラクタ内で `installerOptions.directInstall: true` を設定すると、デフォルトのウェブページを描画する代わりに、ユーザーを直接 Slack の authorize URL に誘導することができます（[例](https://github.com/slackapi/bolt-js/blob/5b4d9ceb65e6bf5cf29dfa58268ea248e5466bfb/examples/oauth/app.js#L58-L64)）。
```

> * **Slack へ追加**: `Add to Slack` ボタンを押すと Slack との OAuth プロセスを開始します。ユーザーがアプリへの権限付与を許可した後、 Slack はアプリの **Redirect URI** （既に設定済み）を呼び出し, ユーザーに **Open Slack** を促します。詳細に設定を変えたい場合は **Redirect URI** セクションを参照してください。

Reviewer: "Slack will call your app’s Redirect URI" can sound a bit unclear as it's actually redirection. I will update the English one later on.

```suggestion
* **Add to Slack (Slack へ追加)**: `Add to Slack` ボタンを押すと Slack との OAuth プロセスを開始します。ユーザーがアプリへの権限付与を許可すると、Slack はアプリの **Redirect URI** （あらかじめ設定されています）へユーザーを誘導し、処理が正常に完了したらユーザーに **Slack で開く** よう促します。これらの設定をカスタマイズする方法については、後述の **Redirect URI** セクションを参照してください。
```

> * **Slack で開く**: ユーザーが **Open Slack** した後、 and here after as your app processes events from Slack, your provided `installationStore`'s `fetchInstallation` と `storeInstallation` ハンドラーが実行されます。ハンドラーに渡す引数についてもっと詳細が知りたい場合、 **Installation Object** セクションを参照してください。

Reviewer: ```suggestion
* **Slack で開く**: ユーザーが **Slack で開く** を選択した後、アプリが Slack からのイベントをするときに `installationStore` の `fetchInstallation` や `storeInstallation` ハンドラーが実行されます。ハンドラーに渡す引数に関するより詳しい情報は  **Installation Object** セクションを参照してください。
```

> * アプリはすでにインストールされていて、さらにユーザーから追加の認可情報（例：ユーザートークンの発行）な場合や、何らかの理由で動的にインストール用の URL を生成したい場合は、`ExpressReceiver` を自前でインスタンス化し、それを `receiver` という変数に代入した上で `receiver.installer.generateInstallUrl()` を呼び出してください。詳細は [OAuth ライブラリのドキュメント](https://slack.dev/node-slack-sdk/oauth#generating-an-installation-url)の `generateInstallUrl()` を参照してください。

Reviewer: ```suggestion
* アプリがすでにインストールされていて、さらにユーザーから追加の認可情報（例：ユーザートークンの発行）な場合や、何らかの理由で動的にインストール用の URL を生成したい場合は、`ExpressReceiver` を自前でインスタンス化し、それを `receiver` という変数に代入した上で `receiver.installer.generateInstallUrl()` を呼び出してください。詳しくは [OAuth ライブラリのドキュメント](https://slack.dev/node-slack-sdk/oauth#generating-an-installation-url)の `generateInstallUrl()` を参照してください。
```

> * 💡 Bolt for JavaScript は [カスタムのレシーバー](#receiver) による OAuth をサポートしていません。もしカスタムのレシーバーを実装したいのであれば、私たちが提供している [OAuth ライブラリ](https://slack.dev/node-slack-sdk/oauth#slack-oauth) を使うこともできます。Bolt for JavaScript の組み込みのモジュールもこれを内部的に利用しています。

Reviewer: ```suggestion
* 💡 Bolt for JavaScript は [カスタムのレシーバー](#receiver)での OAuth をサポートしていません。カスタムのレシーバーで OAuth フローを実装したい場合は、私たちが提供している [OAuth ライブラリ](https://slack.dev/node-slack-sdk/oauth#slack-oauth) を使うことができます。Bolt for JavaScript の組み込みのモジュールもこれを内部的に利用しています。
```

##### Redirect URI

> Bolt for JavaScript は、アプリのインストールフローを完了した後の遷移先の URL である **Redirect URL** のためのパスとして `slack/oauth_redirect` を有効にします。

Reviewer: ```suggestion
Bolt for JavaScript は、アプリのインストールフローを完了した後の遷移先の URL である **Redirect URI** のためのパスとして `/slack/oauth_redirect` を有効にします。
```

💡 アプリのドメインを含んだ **Redirect URI** （絶対 URI）を Slack アプリの設定画面の **OAuth and Permissions** セクション内で設定してください。（例 `https://example.com/slack/oauth_redirect` ）。

> 独自の **Redirect URI** を使う場合、 App クラスの引数 `redirectUri` と `installerOptions.redirectUriPath` にも設定してください。 両方とも設定する必要があり、矛盾のない絶対 URI である必要があります。

Reviewer: ```suggestion
カスタムの **Redirect URI** を使う場合、 App クラスの引数 `redirectUri` と `installerOptions.redirectUriPath` にも設定してください。 両方とも設定する必要があり、また、矛盾のないフル URI である必要があります。
```

```javascript
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  clientId: process.env.SLACK_CLIENT_ID,
  clientSecret: process.env.SLACK_CLIENT_SECRET,
  stateSecret: 'my-state-secret',
  scopes: ['chat:write'],

>   redirectUri: 'https://example.com/slack/redirect', // ココ

Reviewer: ```suggestion
  redirectUri: 'https://example.com/slack/redirect', // ここに設定します
```

  installerOptions: {

>     redirectUriPath: '/slack/redirect', // そして、ココ！

Reviewer: ```suggestion
    redirectUriPath: '/slack/redirect', // ここにも！
```

  },
});
```

##### Installation オブジェクト

> Bolt は `installationStore` の `storeInstallation` ハンドラーに `installation` オブジェクトを渡します。開発者にはオブジェクトの定義が想像しがたく不明瞭かもしれないので、 `installation` オブジェクトのサンプルを載せます:

Reviewer: ```suggestion
Bolt は `installationStore` の `storeInstallation` ハンドラーに `installation` オブジェクトを渡します。どのようなオブジェクトの形式となるか想像しづらいと開発時に混乱の元になるかもしれません。`installation` オブジェクトはこのような形式となります：
```

```javascript
{
  team: { id: 'T012345678', name: 'example-team-name' },
  enterprise: undefined,
  user: { token: undefined, scopes: undefined, id: 'U01234567' },
  tokenType: 'bot',
  isEnterpriseInstall: false,
  appId: 'A01234567',
  authVersion: 'v2',
  bot: {
    scopes: [
      'chat:write',
    ],
    token: 'xoxb-244493-28*********-********************',
    userId: 'U012345678',
    id: 'B01234567'
  }
}
```

> Bolt は `fetchInstallation` と `deleteInstallation` ハンドラーに `installQuery` オブジェクトを渡します:

Reviewer: ```suggestion
Bolt は `fetchInstallation` と `deleteInstallation` ハンドラーに `installQuery` オブジェクトを渡します：
```

```javascript
{
  userId: 'U012345678',
  isEnterpriseInstall: false,
  teamId: 'T012345678',
  enterpriseId: undefined,
  conversationId: 'D02345678'
}
```

##### OrG 全体へのインストール

[Enterprise Grid の OrG 全体へのインストール](https://api.slack.com/enterprise/apps)への対応を追加する場合、Bolt for JavaScript のバージョン 3.0.0 以上を利用してください。また Slack アプリの設定画面で **Org Level Apps** の設定が有効になっていることを確認してください。

> 管理者画面からの [Enterprise Grid の OrG 全体へのインストール](https://api.slack.com/enterprise/apps) の場合、 Bolt で動作させようとすると追加設定が必要です。この場合、推奨されている `state` パラメータが提供されません。 Bolt では `state` による認証を実行しようとし、インストールが止まります。

Reviewer: ```suggestion
管理者画面からの [Enterprise Grid の OrG 全体へのインストール](https://api.slack.com/enterprise/apps) の場合、 Bolt で動作させるために追加の設定が必要です。この利用シナリオでは、推奨の `state` パラメータが提供されず、Bolt アプリでは `state` を検証しようとするため、インストールを継続することができません。
```

> Bolt では、App の `stateVerification` オプションを false にすることで、 state による認証を無効にできます。以下の例を参考にしてください。

Reviewer: ```suggestion
Bolt アプリ側で `stateVerification` オプションを false に設定することで、 `state` パラメーターの検証を無効することができます。以下の例を参考にしてください。
```

```javascript
const app = new App({
  signingSecret: process.env.SLACK_SIGNING_SECRET,
  clientId: process.env.SLACK_CLIENT_ID,
  clientSecret: process.env.SLACK_CLIENT_SECRET,
  scopes: ['chat:write'],
  installerOptions: {
    stateVerification: false,
  },
});
```

> Slack の OAuth インストールフローについてもっと知りたい場合は [API ドキュメント](https://api.slack.com/authentication/oauth-v2)を参照してください。

Reviewer: ```suggestion
Slack の OAuth インストールフローについてのより詳細な情報は [API ドキュメント](https://api.slack.com/authentication/oauth-v2)を参照してください。
```
